### DISCUSSION TITLE
Trouble with a deny-by-default authorization policy across namespaces with server-first protocols

### DISCUSSION THREAD

**Discussion by AlesKrajnik** [istio/istio#48524] [Q&A] [answered]
Hi all,

I am trying to implement a deny-by-default authorization policy, but it seems not to be working well across different namespaces.

In my example I use the following names:

* namespace `targetNS` with peer authentication mTLS mode `STRICT`
  * deployment `targetDeployA`, labeled `app.kubernetes.io/name: targetDeployA`, running under service account `targetAccountA`
  * deployment `targetDeployB`, running under service account `targetAccountB`
* namespace `sourceNS` with peer authentication mTLS mode `STRICT`
  * deployment `sourceDeploy`, running under service account `sourceAccount`

In my `targetNS` namespace, I have set up this:

```yaml
apiVersion: security.istio.io/v1
kind: AuthorizationPolicy
metadata:
  name: default
  namespace: targetNS
spec:
  action: ALLOW
```

According to the documentation, such authorization policy matches all workloads and allows nothing, so it should be effectively a deny-by-default policy.

Now I created an explicit `ALLOW` policy for `targetDeployA` workload, listing all principals allowed to connect, like this:

```yaml
apiVersion: security.istio.io/v1
kind: AuthorizationPolicy
metadata:
  name: targetAuthorizationPolicyA
  namespace: targetNS
spec:
  action: ALLOW
  selector:
    matchLabels:
      app.kubernetes.io/name: targetDeployA
  rules:
    - to:
        - operation:
            ports:
              - "25"
      from:
        - source:
            principals:
              - "cluster.local/ns/targetNS/sa/targetAccountB"
              - "cluster.local/ns/sourceNS/sa/sourceAccount"
```

The peer authentication in both namespaces is set up like this:

```yaml
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: default
  namespace: targetNS
spec:
  mtls:
    mode: STRICT
```

Prior to creating `targetAuthorizationPolicyA`, `targetDeployB` could not connect, when I created the `targetAuthorizationPolicyA`, the `targetDeployB` can connect. If I remove the `targetAccountB` principal from the `targetAuthorizationPolicyA` policy (or remove the policy completely), the `targetDeployB` can no longer connect. As expected.

However, the `sourceDeploy` can always connect, indendepently of whether the `targetAuthorizationPolicyA` exists or not and whether the principal `cluster.local/ns/sourceNS/sa/sourceAccount` is listed there or not.

It seems that when the `targetAuthorizationPolicyA` exists as written above, the connection matches and allows the `sourceDeploy` to talk to `targetDeployA`. When I remove the respective principal or delete the whole policy `targetAuthorizationPolicyA`, obviously the match does not happen, however the default does not apply - so the traffic is allowed with `rbac_access_denied_matched_policy[none]` logged in the `istio-proxy` output.

It seems that the first `default` policy applies correctly to `targetNS` namespace, but does not apply to `sourceNS`? I also tried to create similar `default` policy in the `sourceNS`, but that did not change the situation.

What am I missing?

Thank you!

**Comment by rootsongjc** [comment]
Based on your description, it seems like you want to implement the following authorization policy:

```mermaid
graph TD
    subgraph sourceNS
        sourceDeploy["sourceDeploy (Service Account: sourceAccount)"]
    end

    subgraph targetNS
        targetDeployA["targetDeployA (Service Account: targetAccountA)"]
        targetDeployB["targetDeployB (Service Account: targetAccountB)"]
    end

    defaultPolicy["Default Deny Policy (targetNS)"]
    targetAuthorizationPolicyA["AuthorizationPolicyA (Allow targetDeployA)"]

    sourceDeploy -->|Allowed/Denied| targetDeployA
    sourceDeploy -->|Denied| targetDeployB
    targetDeployB -->|Allowed by targetAuthorizationPolicyA| targetDeployA

    classDef namespace fill:#ddd,stroke:#333,stroke-width:1px;
    classDef service fill:#bdf,stroke:#036,stroke-width:1px;
    classDef policy fill:#faa,stroke:#a33,stroke-width:1px;
    class sourceDeploy,targetDeployA,targetDeployB service;
    class targetNS,sourceNS namespace;
    class defaultPolicy,targetAuthorizationPolicyA policy;
```

Only `sourceDeploy` in `sourceNS` and `targetDeployB` are allowed to access `targetDeployA`, while other services in both `sourceNS` and `targetNS` are denied access to it.

Your `default` AuthorizationPolicy should be set like this:

```yaml
apiVersion: security.istio.io/v1
kind: AuthorizationPolicy
metadata:
  name: default
  namespace: targetNS
spec:
  {}
```

This policy ensures that traffic to all services in `targetNS` is denied by default when no other specific policies match.

Your `targetAuthorizationPolicyA` configuration looks correct.

PS. The decision logic for Istio's authorization policies is provided below for your reference:

```mermaid
graph TD
    A[Start] --> B{Any CUSTOM policies match?}
    B -- Yes --> C{Do CUSTOM policies evaluate to deny?}
    C -- Yes --> D[Deny the request]
    C -- No --> E{Any DENY policies match?}
    B -- No --> E
    E -- Yes --> D
    E -- No --> F{Any ALLOW policies for the workload?}
    F -- No --> G[Allow the request]
    F -- Yes --> H{Any ALLOW policies match?}
    H -- Yes --> G
    H -- No --> D
```

**Comment by AlesKrajnik** [comment]
Hi @rootsongjc,

thank you! What you depicted there, is exactly what I am trying to implement.

Your suggestion how to define the `default` authorization policy, however, yielded no result. That is because your policy is effectively the same as mine. While I use

```yaml
spec:
  action: ALLOW
```

You suggested

```yaml
spec: {}
```

Since `ALLOW` is the default ([documentation](https://istio.io/v1.20/docs/reference/config/security/authorization-policy/#AuthorizationPolicy-Action)), the policies are equivalent.

However, I found out that my `default` policy was actually properly blocking the traffic from `sourceNS` as expected, so everything worked correctly. What confused me is that I tested with SMTP protocol (25), which is a [server-first protocol](https://istio.io/v1.20/docs/ops/deployment/requirements/#server-first-protocols). Apparently, the authorization policy does not stop the server in the case of a server-first protocol to send its initial message and only when the client responds, only then the authorization policy kicks in and shuts down the traffic.

Therefore, when I was testing the policy, I only tried to establish connection to `targetDeployA` and that always worked, because the server always sent the initial greeting:

```
Trying 10.245.168.122...
Connected to targetDeployA.targetNS.
Escape character is '^]'.
220 targetDeployA ESMTP Postfix
```

From this I concluded that the authorization policy did not work because it allowed the traffic. But the difference was actually what happened afterwards.

With only the `default` policy in place, which was denying all traffic by default, this is what followed after the client tried to talk back to the server after the initial greeting:

```
Trying 10.245.168.122...
Connected to targetDeployA.targetNS.
Escape character is '^]'.
220 targetDeployA ESMTP Postfix
HELO localhost
Connection closed by foreign host.
```

So istio killed the communication correctly. With `targetAuthorizationPolicyA` in place, allowing `sourceAccount` to connect, this is what happened:

```
Trying 10.245.168.122...
Connected to targetDeployA.targetNS.
Escape character is '^]'.
220 targetDeployA ESMTP Postfix
HELO localhost
250 targetDeployA
```

**So, let me rephrase the question: can an authorization policy stop client even from connecting, that is stop the server from sending an initial greeting in the case of server-first protocols?**

@rootsongjc Thanks for kicking me in the right direction by confirming to me that my setup should work! That made me look further and actually found the issue.

Thank you everyone for helping!

**Reply by rootsongjc** [reply, marked as answer]
I'm glad to help. 😊 I feel they are not on the same level. TCP connections are established at the transport layer independently of Istio's control, while Istio enforces authorization policies at the application layer. You should use NetworkPolicy to achieve this. For example, you can consider Cilium CNI or other CNI plugins. You can refer to this link: https://docs.cilium.io/en/v1.9/policy/language/#layer-4-examples
